### PR TITLE
Correct Implementation for imt

### PR DESCRIPTION
**Pull request by 0xVikasRushi**
### Circuit Changes
1. `compute_merkle_root` - changed logic to handle empty siblings
1. `insert_leaf` - added old root contraint and verify intermediate leaf with zero

### Circuit test cases
1. Added `test_compute_merkle_root`
2. Added `test_verify_merkle_proof`
3. Added  `test_verify_non_inclusion`
4. Updated `test_insert_leaf`
5. Updated `test_test_insert_leaf_multiple_round`

### Native Rust code
1. updated new  `IndexedMerkleTree` implementation instead of zero hashes
2. updated `insert_leaf` according to spec
3. added tests for insert_leaf and merkle root calculation as well.

### CODE REVIEW

File: src/indexed_merkle_tree.rs [hunk 1]
```diff
@@ -262,16 +256,20 @@ pub fn insert_leaf<F: BigPrimeField, const T: usize, const RATE: usize>(
...
         gate,
-        &[newlowleaf.val, newlowleaf.next_val, newlowleaf.next_idx],
+        &[
+            new_low_leaf.val,
+            new_low_leaf.next_val,
+            new_low_leaf.next_idx,
+        ],
     );
 
     let interim_root = compute_merkle_root(
```
Comment (Vishalkulkarni45) [inline src/indexed_merkle_tree.rs:275]: why have you replaced the `zero_leaf_hash`, `new_leaf_proof` , `new_leaf_proof_helper` with  `new_low_leaf_hash`,`low_leaf_proof`,`low_leaf_proof_helper` ? 
Can you once revisit the logic and let me know what are changes are to be made 
You need to make changes in how the parent hash is calculated using the child nodes and not in how the intermediate roots are calculated

**Review by Vishalkulkarni45 — CHANGES_REQUESTED** [review]
Thanks @0xVikasRushi for the PR.  You have changed the logic of calculating the intermediate root where you calculate it by replacing the  `zero-hash-leaf` with the `new_low_leaf_hash` and using the same old siblings . In the next step you verify it which does't make sense to me because you have calculated the `intermediate root` inside the circuit and again you are verifying the same computation inside  circuit.

The bug in the circuit is prover can provide the low leaf and its sibling of `IMT-1` and provide the new_leaf and its sibling of `IMT-2` and still the circuit will provide the prover a valid proof .

In your current logic there is no integrity between the low leaf and the new leaf to be inserted .

I suggest you to first only look at how to make changes in the calculation of the parent hash , look into the halo2-scroll how i have done it .

File: src/indexed_merkle_tree.rs [hunk 2]
```diff
@@ -288,9 +286,9 @@ pub fn insert_leaf<F: BigPrimeField, const T: usize, const RATE: usize>(
         range,
         hasher,
         &interim_root,
-        &zero_leaf_hash,
-        new_leaf_proof,
-        new_leaf_proof_helper,
+        &new_low_leaf_hash,
+        low_leaf_proof,
+        low_leaf_proof_helper,
```
Comment (Vishalkulkarni45) [inline src/indexed_merkle_tree.rs]: why have you replaced the zero_leaf_hash, new_leaf_proof , new_leaf_proof_helper with new_low_leaf_hash,low_leaf_proof,low_leaf_proof_helper ?
Can you once revisit the logic and let me know what are changes are to be made
You need to make changes in how the parent hash is calculated using the child nodes and not in how the intermediate roots are calculated

**Review by Vishalkulkarni45 — CHANGES_REQUESTED** [review]

File: src/indexed_merkle_tree.rs [hunk 3]
```diff
@@ -262,16 +256,20 @@ pub fn insert_leaf<F: BigPrimeField, const T: usize, const RATE: usize>(
         is_new_leaf_largest,
```
Comment (rahulghangas) [inline src/indexed_merkle_tree.rs]: Let's replace this as well. There is no need to pass `is_new_leaf_largest` as external witness. Instead we can create two (in)equalities. 
```
low_leaf.next > new_leaf
```
and 
```
low_leaf.next == 0
```

then use an `OR` gate for both

**Review by rahulghangas — CHANGES_REQUESTED** [review]

File: src/indexed_merkle_tree.rs (hunk 3 above)
Comment (Vishalkulkarni45) [reply src/indexed_merkle_tree.rs]: @0xVikasRushi fix this also

File: src/indexed_merkle_tree.rs [hunk 4]
```diff
@@ -208,17 +207,8 @@ pub fn verify_non_inclusion<F: BigPrimeField, const T: usize, const RATE: usize>
...
-        *one,
-        *is_new_leaf_largest,
-        is_zero,
-        is_next_val_greater,
-    );
-
-    assert_eq!(is_true.value(), &F::ONE);
-    ctx.constrain_equal(&is_true, &one);
+    let condition = gate.or(ctx, is_zero, is_next_val_greater);
+    ctx.constrain_equal(&condition, one);
```
Comment (0xVikasRushi) [inline src/indexed_merkle_tree.rs:211]: 1.  `is_new_leaf_greater` = new_leaf < low_leaf.next
2. `is_zero`  = low_leaf.next_val === 0

constrain **OR(is_zero, is_new_leaf_greater)**

File: src/indexed_merkle_tree.rs (hunk 3 above)
Comment (0xVikasRushi) [reply src/indexed_merkle_tree.rs]: Updated [commit](https://github.com/aerius-labs/indexed-merkle-tree-halo2/pull/1/commits/67f4708319b72bcb083f581324669e2570f020f2#r1760205321)

**Review by Vishalkulkarni45 — APPROVED** [review]

**Review by rahulghangas — APPROVED** [review]

**Pull request merged**